### PR TITLE
Fixing ShowInTaskbar=false doesn't work in some cases (#6421)

### PR DESCRIPTION
**Pull request by roland5572**
Fixes  https://github.com/dotnet/winforms/issues/6421.

If you set ShowInTaskbar=false, the winforms will recreate the Handle. It will call RecreateHandleCore() from Control.cs.

In method WmNCDestroy(ref Message m) from class Form.cs, The program needs to confirm whether the handle is being recreated before setting the value of the DialogResult.

If the program is in the process of recreating the handle, setting DialogResult = DialogResult.Cancel will cause the Form to be automatically closed.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/6989)

### CODE REVIEW

**Comment by Tanya-Solyanik** [comment]
Looks good, could you please add unit test for this case

**Comment by roland5572** [comment]
@Tanya-Solyanik Thank you. I have tested it on my machine. I am new to this, do you have a spec document so I can learn how to make unit tests that conform to the .net spec? Because this change actually only adjusts one line of the statement. I'm not sure about the test system's rules for dealing with such changes.

**Comment by Tanya-Solyanik** [comment]
@roland5572  - you could use this https://github.com/dotnet/winforms/blob/main/docs/testing.md as a starting point.

**Comment by roland5572** [comment]
@Tanya-Solyanik Thank you. Do you prefer a unit test for current issue, or a unit test specifically tests a function that has been fixed?

**Comment by Tanya-Solyanik** [comment]
>Do you prefer a unit test for current issue, or a unit test specifically tests a function that has been fixed?

Ha-ha, may I have both please?

I meant the current issue. We are aiming to accompany fixes with "regression" tests that cover specific issues being fixed. But if you have time to cover this functionality, such a PR will always be welcome.

**Comment by RussKie** [comment]
I'm currently OOF, and unable to review. Just want to note that if tests need to show forms, those tests should be thought of as integration tests. And the best place for those is UIIntegrationTests project. Before adding an automated test please start with a manual test, it should be added to WinFormsControlTests project (writing project names from memory).

One of the best ways to approach a complex issue is to write a test (and have it failed) in the first commit, then add a fix in a second commit (at this point the test must pass). This approach works great if you're new to testing as well.
________________________________

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=6989) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=6989) before we can accept your contribution.<br/><br/>:x: roland5572 [sign now](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=6989)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/winforms?pullRequest=6989) it.</sub>

**Comment by roland5572** [comment]
Thank you, I have added unit test.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs [hunk 1]
```diff
@@ -1820,6 +1820,32 @@ public void Form_TransparencyKey_SetNotTopLevelWithHandle_GetReturnsExpected(boo
             Assert.Equal(0, createdCallCount);
         }
 
+        [ActiveIssue("https://github.com/dotnet/winforms/issues/6421")]
+        [WinFormsFact]
+        public void Form_ShowInTaskBar_SetFalse_GetReturnsExpected()
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1825]: nit: `Taskbar`  - B is not capitalized

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs [hunk 2]
```diff
@@ -1820,6 +1820,32 @@ public void Form_TransparencyKey_SetNotTopLevelWithHandle_GetReturnsExpected(boo
...
+            using var control = new Form
+            {
+                ShowInTaskbar = true,
+            };
+
+            var prefValue = DialogResult.OK;
+
+            control.Load += (object sender, EventArgs e) =>
+            {
+                control.ShowInTaskbar = false;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1836]: Is it possible to confirm that handle had been re-created? Perhaps compare handle values? Or read the state?

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs [hunk 3]
```diff
@@ -1820,6 +1820,32 @@ public void Form_TransparencyKey_SetNotTopLevelWithHandle_GetReturnsExpected(boo
...
+            };
+
+            control.Shown += (object sender, EventArgs e) =>
+            {
+                control.DialogResult = prefValue;
+            };
+
+            control.ShowDialog();
+
+            Assert.Equal(prefValue, control.DialogResult);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1846]: Please confirm that the handle had been created.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs [hunk 4]
```diff
@@ -1820,6 +1820,32 @@ public void Form_TransparencyKey_SetNotTopLevelWithHandle_GetReturnsExpected(boo
...
+            {
+                control.ShowInTaskbar = false;
+            };
+
+            control.Shown += (object sender, EventArgs e) =>
+            {
+                control.DialogResult = prefValue;
+            };
+
+            control.ShowDialog();
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1844]: ShowDialog should return the same value, you could verify that as well.

**Review by Tanya-Solyanik — COMMENTED** [review]
Thank you for adding a test! I added some nits

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs [hunk 5]
```diff
@@ -1820,6 +1820,32 @@ public void Form_TransparencyKey_SetNotTopLevelWithHandle_GetReturnsExpected(boo
...
+        [ActiveIssue("https://github.com/dotnet/winforms/issues/6421")]
+        [WinFormsFact]
+        public void Form_ShowInTaskBar_SetFalse_GetReturnsExpected()
+        {
+            using var control = new Form
+            {
+                ShowInTaskbar = true,
+            };
+
+            var prefValue = DialogResult.OK;
```
[the remark this answers is not in the document]
Comment (roland5572) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1832]: I think it is preference value. If you prefer "expected", I'll use expected all in the future.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs [hunk 6]
```diff
@@ -1820,6 +1820,32 @@ public void Form_TransparencyKey_SetNotTopLevelWithHandle_GetReturnsExpected(boo
             Assert.Equal(0, createdCallCount);
         }
 
+        [ActiveIssue("https://github.com/dotnet/winforms/issues/6421")]
+        [WinFormsFact]
+        public void Form_ShowInTaskBar_SetFalse_GetReturnsExpected()
+        {
+            using var control = new Form
```
[the remark this answers is not in the document]
Comment (roland5572) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1827]: I can see many "using var control = new Form" from FormTests.cs, So I'm not sure which is better. I'll use form in the future.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs (hunk 4 above)
Comment (roland5572) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1844]: Assert.Equal(prefValue, control.DialogResult) has implemented the check. My thinking is to check the property at the end is also correct. Do you tend to check both?

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1832]: Code style guide recommends against using abbreviations. The rest is up to you.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs (hunk 2 above)
Comment (roland5572) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1836]: In this Issue, Handle is bound to be created successfully. This Issue is not caused by the failure to create the Handle, but in the process of creating the Handle, a statement calls the statement that automatically closes the window.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs (hunk 4 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1844]: We expect them to match, so makes sense to check both

**Comment by roland5572** [comment]
@Tanya-Solyanik Thank you. I have modified all the code according to your suggestion, and added a manual test.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs [hunk 7]
```diff
@@ -1820,6 +1820,37 @@ public void Form_TransparencyKey_SetNotTopLevelWithHandle_GetReturnsExpected(boo
...
+
+            bool expectedIsHandleCreated = true;
+            DialogResult expectedDialogResult = DialogResult.OK;
+
+            form.Load += (object sender, EventArgs e) =>
+            {
+                IntPtr formHandle = form.Handle;
+                form.ShowInTaskbar = false;
+
+                Assert.Equal(expectedIsHandleCreated, form.IsHandleCreated);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/FormTests.cs:1841]: nit: `Assert.True(form.IsHandleCreated)`

**Review by Tanya-Solyanik — DISMISSED** [review]
Thank you, Looks good

**Review by Tanya-Solyanik — APPROVED** [review]
Thank you!

**Comment by AraHaan** [comment]
Will this go into a servicing release for .NET 6?

**Comment by dreddy-work** [comment]
@AraHaan , given this has been the behavior since .NET framework, it wouldn't meet the servicing bar unless there is a business justification on customer impact.

**Comment by AraHaan** [comment]
So basically winforms programs targeting .NET 6 (and might be production quality) will still be broken by this as it won't be serviced?

**Comment by RussKie** [comment]
> So basically winforms programs targeting .NET 6 (and might be production quality) will still be broken by this as it won't be serviced as they cannot target .NET 7 yet until RC at least?

Since this was an existing issue apps targeting .NET 6.0 aren't affected by this issue any more than apps targeting .NET Framework.

**Pull request merged**